### PR TITLE
Track transform

### PR DESCRIPTION
**Pull request by izzytwosheds**
Implementation of track based transformations, such as ability to mux/demux tracks.

### CODE REVIEW

File: litr-demo/src/main/java/com/linkedin/android/litr/demo/data/AudioTarget.java [hunk 1]
```diff
@@ -22,6 +23,7 @@ public AudioTarget() {
     @Override
     public void onCheckedChanged(CompoundButton button, boolean checked) {
         shouldTranscodeAudio = checked;
+        shouldKeepTrack = checked;
```
Comment (lopeng) [inline litr-demo/src/main/java/com/linkedin/android/litr/demo/data/AudioTarget.java]: is it confusing? e.g. when shouldTranscodeAudio is false, then shouldKeepTrack is false

File: litr-demo/src/main/java/com/linkedin/android/litr/demo/data/TransformationPresenter.java [hunk 2]
```diff
@@ -54,26 +70,105 @@ public void startTransformation(@NonNull SourceMedia sourceMedia,
...
+            throw new IllegalArgumentException("No output tracks!");
+        }
+
+        try {
+            MediaTarget mediaTarget = new MediaMuxerMediaTarget(transformationState.targetFile.getAbsolutePath(),
+                                                                trackCount,
+                                                                sourceMedia.videoRotation,
+                                                                MediaMuxer.OutputFormat.MUXER_OUTPUT_MPEG_4);
+
+            List<TrackTransform> trackTransforms = new ArrayList<>(2);
```
Comment (lopeng) [inline litr-demo/src/main/java/com/linkedin/android/litr/demo/data/TransformationPresenter.java]: new ArrayList<>(2) or new ArrayList<>(trackCount)?

File: litr-demo/src/main/java/com/linkedin/android/litr/demo/data/AudioTarget.java (hunk 1 above)
Comment (lopeng) [inline litr-demo/src/main/java/com/linkedin/android/litr/demo/data/AudioTarget.java]: While shouldTranscodeAudio true then shouldKeepTrack true sounds right, shouldTranscodeAudio false then shouldKeepTrack false seems confusing.

File: litr-demo/src/main/java/com/linkedin/android/litr/demo/data/TransformationPresenter.java [hunk 3]
```diff
@@ -54,26 +70,105 @@ public void startTransformation(@NonNull SourceMedia sourceMedia,
...
+                                       sourceMedia.uri,
+                                       transformationState.targetFile.getAbsolutePath(),
+                                       createMediaFormat(sourceMedia, videoTarget),
+                                       createMediaFormat(sourceMedia, audioTarget),
+                                       transformationListener,
+                                       MediaTransformer.GRANULARITY_DEFAULT,
+                                       createGlFilters(sourceMedia, overlayTarget));
+        } else {
+            performTrackTransformation(sourceMedia, videoTarget, audioTarget, overlayTarget, transformationState, transformationListener);
         }
```
Comment (lopeng) [inline litr-demo/src/main/java/com/linkedin/android/litr/demo/data/TransformationPresenter.java]: Can performTrackTransformation logic handle both cases?

File: litr/src/main/java/com/linkedin/android/litr/TrackTransform.java [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        private final MediaTarget mediaTarget;
+
+        private Decoder decoder;
+        private VideoRenderer renderer;
+        private Encoder encoder;
+        private MediaFormat targetFormat;
+        private int targetTrack;
+
+        public Builder(@NonNull MediaSource mediaSource,
+                       int sourceTrack,
```
Comment (lopeng) [inline litr/src/main/java/com/linkedin/android/litr/TrackTransform.java]: add @IntRange?

File: litr/src/main/java/com/linkedin/android/litr/utils/TranscoderUtils.java [hunk 5]
```diff
@@ -29,51 +30,40 @@
...
-
-    public static long getEstimatedTargetVideoFileSize(@NonNull MediaSource mediaSource,
-                                                       @NonNull MediaFormat targetVideoFormat,
-                                                       @Nullable MediaFormat targetAudioFormat) {
+    private TranscoderUtils() {}
+
+    /**
+     * Estimate total target file size(s) for track transformation
+     * @param trackTransforms track transformations
+     * @return estimated size, zero if estimation fails
```
Comment (lopeng) [inline litr/src/main/java/com/linkedin/android/litr/utils/TranscoderUtils.java]: 0 vs -1?

File: litr-demo/src/main/java/com/linkedin/android/litr/demo/data/TransformationPresenter.java (hunk 2 above)
Comment (izzytwosheds) [reply litr-demo/src/main/java/com/linkedin/android/litr/demo/data/TransformationPresenter.java]: 2 is a maximum. 2 vs. 1 is not a big hit memory wise, and saves us from doing more complex logic to calculate it.

File: litr-demo/src/main/java/com/linkedin/android/litr/demo/data/AudioTarget.java (hunk 1 above)
Comment (izzytwosheds) [reply litr-demo/src/main/java/com/linkedin/android/litr/demo/data/AudioTarget.java]: I think shouldTranscodeAudio=true && shouldKeepTrack=false sounds even more ridiculous (transcode and throw away)
We should come up with a better API for this. Probably something that will more closely mimic MediaTransformer API. For now, I don't care about this too much, but soon sample app has to be refactored to have more flexible data classes.

File: litr/src/main/java/com/linkedin/android/litr/utils/TranscoderUtils.java (hunk 5 above)
Comment (izzytwosheds) [reply litr/src/main/java/com/linkedin/android/litr/utils/TranscoderUtils.java]: Yeah, -1 is better

File: litr-demo/src/main/java/com/linkedin/android/litr/demo/data/AudioTarget.java (hunk 1 above)
Comment (lopeng) [reply litr-demo/src/main/java/com/linkedin/android/litr/demo/data/AudioTarget.java]: Need a better API otherwise it's confusing.

**Pull request merged**